File: lib/disease-risk/aggregate.ts
/**
 * Per-cell disease assessment: hourly weather -> daily stats -> risk band.
 * Faithful TypeScript port of Prahari engine/aggregate.py (pure, no I/O).
 *
 *   * Hutton daily gate uses TOTAL wet hours in the day and daily min temp.
 *   * The Hutton criterion needs those qualifying days to be CONSECUTIVE.
 *   * Wallin DSV uses the LONGEST wet SPELL and the mean temperature DURING that spell.
 *   * Bands come from accumulated DSV thresholds (spray -> act, amber -> watch).
 */
import { criterionMet, hoursRhAtOrAbove } from './hutton'
import { wallinDsv } from './wallin'
import type { HuttonParams, SeverityConfig } from './models'

export type Band = 'safe' | 'watch' | 'act'

export interface DayStats {
  date: string
  /** total hours RH >= threshold (Hutton 6 h gate) */
  wetHours: number
  /** longest consecutive wet spell (drives DSV) */
  spellHours: number
  minTempC: number
  /** mean temp DURING the longest wet spell */
  meanWetTempC: number
  precipMm: number
  /** Hutton daily gate: minTemp >= minTempC AND wetHours >= minWetHours */
  qualifies: boolean
  dsv: number
}

export interface CellAssessment {
  band: Band
  risk: number
  physicsRisk: number
  /** 0.0 — physics only, no ML correction (honest: there is no trained model here). */
  mlDelta: number
  criterionMet: boolean
  dsvToday: number
  dsvAccum: number
  wetHours: number
  minTempC: number
  meanWetTempC: number
}

function round(value: number, digits: number): number {
  const f = 10 ** digits
  return Math.round(value * f) / f
}

/** Group hourly indices by calendar day using the "YYYY-MM-DD" prefix (pure string work). */
function dailyIndexGroups(times: readonly string[]): [string, number[]][] {
  const groups = new Map<string, number[]>()
  const order: string[] = []
  times.forEach((t, idx) => {
    const day = t.slice(0, 10)
    if (!groups.has(day)) {
      groups.set(day, [])
      order.push(day)
    }
    groups.get(day)!.push(idx)
  })
  return order.map((day) => [day, groups.get(day)!])
}

export function dailyStats(
  times: readonly string[],
  temperature: readonly number[],
  relativeHumidity: readonly number[],
  precipitation: readonly number[],
  params: HuttonParams,
  severity: SeverityConfig,
): DayStats[] {
  const rhThr = params.rhThreshold
  const minWet = params.minWetHours
  const minT = params.minTempC

  const out: DayStats[] = []
  for (const [day, idxs] of dailyIndexGroups(times)) {
    const dayRh = idxs.map((i) => relativeHumidity[i])
    const dayTemp = idxs.map((i) => temperature[i])

    const wetHours = hoursRhAtOrAbove(dayRh, rhThr)

    // Longest wet spell and the mean temperature during it.
    let bestLen = 0
    let bestSum = 0.0
    let curLen = 0
    let curSum = 0.0
    for (let k = 0; k < dayRh.length; k++) {
      if (dayRh[k] >= rhThr) {
        curLen += 1
        curSum += dayTemp[k]
        if (curLen > bestLen) {
          bestLen = curLen
          bestSum = curSum
        }
      } else {
        curLen = 0
        curSum = 0.0
      }
    }
    const meanWetTemp = bestLen ? bestSum / bestLen : 0.0

    const minTemp = dayTemp.length ? Math.min(...dayTemp) : 0.0
    const precip = idxs.reduce((s, i) => s + precipitation[i], 0.0)
    const qualifies = minTemp >= minT && wetHours >= minWet
    const dsv = wallinDsv(meanWetTemp, bestLen, severity.dsvTable)

    out.push({
      date: day,
      wetHours,
      spellHours: bestLen,
      minTempC: round(minTemp, 2),
      meanWetTempC: round(meanWetTemp, 2),
      precipMm: round(precip, 2),
      qualifies,
      dsv,
    })
  }
  return out
}

/** Combine daily stats into a risk band. mlDelta is 0 (physics only). */
export function assessCell(
  days: readonly DayStats[],
  params: HuttonParams,
  severity: SeverityConfig,
): CellAssessment {
  const crit = criterionMet(days.map((d) => d.qualifies), params.consecutiveDays)
  const dsvAccum = days.reduce((s, d) => s + d.dsv, 0)
  const dsvToday = days.length ? days[0].dsv : 0

  const spray = severity.sprayThresholdDsv
  const amber = severity.amberThresholdDsv

  // Severity-gated bands: the Hutton criterion escalates to "watch" only when disease severity
  // is actually developing (dsvAccum > 0). Criterion met but DSV 0 (wet-spell temp outside the
  // pathogen's viable range) stays "safe" — avoiding a district-wide false alarm.
  let band: Band
  if (dsvAccum >= spray) band = 'act'
  else if (dsvAccum >= amber || (crit && dsvAccum > 0)) band = 'watch'
  else band = 'safe'

  const physicsRisk = spray ? Math.min(1.0, dsvAccum / spray) : 0.0
  return {
    band,
    risk: round(physicsRisk, 3),
    physicsRisk: round(physicsRisk, 3),
    mlDelta: 0.0,
    criterionMet: crit,
    dsvToday,
    dsvAccum,
    wetHours: days.length ? days[0].wetHours : 0,
    minTempC: days.length ? days[0].minTempC : 0.0,
    meanWetTempC: days.length ? days[0].meanWetTempC : 0.0,
  }
}

export const BAND_RANK: Record<Band | 'unknown', number> = { act: 0, watch: 1, safe: 2, unknown: 3 }

File: PRAHARI_INTELLIGENCE/engine/aggregate.py
"""Per-cell disease assessment: hourly weather -> daily stats -> risk band. Pure.

Ties the rules together (PRD §8.2 Step 3-4):
  * Hutton daily gate uses TOTAL wet hours in the day (hours_rh_at_or_above) and daily min temp.
  * The Hutton criterion needs those qualifying days to be CONSECUTIVE (criterion_met).
  * Wallin DSV uses the LONGEST wet SPELL and the mean temperature DURING that spell.
Bands come from accumulated DSV thresholds in models.yaml (spray -> act, amber -> watch).
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

from engine.rules import criterion_met, hours_rh_at_or_above
from engine.wallin import wallin_dsv


@dataclass(frozen=True)
class DayStats:
    date: str
    wet_hours: int          # total hours RH >= threshold (Hutton 6 h gate)
    spell_hours: int        # longest consecutive wet spell (drives DSV)
    min_temp_c: float
    mean_wet_temp_c: float  # mean temp DURING the longest wet spell
    precip_mm: float
    qualifies: bool         # Hutton daily gate: min_temp >= min_temp_c AND wet_hours >= min_wet_hours
    dsv: int


@dataclass(frozen=True)
class CellAssessment:
    band: str               # safe | watch | act
    risk: float
    physics_risk: float
    ml_delta: float         # 0.0 in Phase 1 (no ML correction yet)
    criterion_met: bool
    dsv_today: int
    dsv_accum: int
    wet_hours: int
    min_temp_c: float
    mean_wet_temp_c: float


def _daily_index_groups(times: Sequence[str]) -> list[tuple[str, list[int]]]:
    groups: dict[str, list[int]] = {}
    order: list[str] = []
    for idx, t in enumerate(times):
        day = t[:10]  # "YYYY-MM-DD" prefix — pure string work, no clock
        if day not in groups:
            groups[day] = []
            order.append(day)
        groups[day].append(idx)
    return [(day, groups[day]) for day in order]


def daily_stats(
    times: Sequence[str],
    temperature: Sequence[float],
    relative_humidity: Sequence[float],
    precipitation: Sequence[float],
    params: dict,
    dsv_table=None,
) -> list[DayStats]:
    rh_thr = params["rh_threshold"]
    min_wet = params["min_wet_hours"]
    min_t = params["min_temp_c"]

    out: list[DayStats] = []
    for day, idxs in _daily_index_groups(times):
        day_rh = [relative_humidity[i] for i in idxs]
        day_temp = [temperature[i] for i in idxs]

        wet_hours = hours_rh_at_or_above(day_rh, rh_thr)

        # Longest wet spell and the mean temperature during it.
        best_len, best_sum = 0, 0.0
        cur_len, cur_sum = 0, 0.0
        for rh, tp in zip(day_rh, day_temp):
            if rh >= rh_thr:
                cur_len += 1
                cur_sum += tp
                if cur_len > best_len:
                    best_len, best_sum = cur_len, cur_sum
            else:
                cur_len, cur_sum = 0, 0.0
        mean_wet_temp = (best_sum / best_len) if best_len else 0.0

        min_temp = min(day_temp) if day_temp else 0.0
        precip = sum(precipitation[i] for i in idxs)
        qualifies = (min_temp >= min_t) and (wet_hours >= min_wet)
        dsv = (wallin_dsv(mean_wet_temp, best_len) if dsv_table is None
               else wallin_dsv(mean_wet_temp, best_len, dsv_table))

        out.append(DayStats(day, wet_hours, best_len, round(min_temp, 2),
                            round(mean_wet_temp, 2), round(precip, 2), qualifies, dsv))
    return out


def assess_cell(days: Sequence[DayStats], params: dict, severity: dict) -> CellAssessment:
    """Combine daily stats into a risk band. ml_delta is 0 in Phase 1 (physics only)."""
    crit = criterion_met([d.qualifies for d in days], params["consecutive_days"])
    dsv_accum = sum(d.dsv for d in days)
    dsv_today = days[0].dsv if days else 0

    spray = severity["spray_threshold_dsv"]
    amber = severity["amber_threshold_dsv"]
    # Severity-gated bands: the Hutton criterion escalates to "watch" only when disease
    # severity is actually developing (dsv_accum > 0). When the criterion is met but DSV is 0
    # because the wet-spell temperature is outside the pathogen's viable range (e.g. > 26.6 C
    # in monsoon heat), the cell is "safe" for this disease — avoiding a district-wide false
    # alarm (PRD §37: false alarms are the worst outcome). mean_wet_temp_c is exposed in the
    # artefact so this reason is auditable.
    if dsv_accum >= spray:
        band = "act"
    elif dsv_accum >= amber or (crit and dsv_accum > 0):
        band = "watch"
    else:
        band = "safe"

    physics_risk = min(1.0, dsv_accum / spray) if spray else 0.0
    return CellAssessment(
        band=band,
        risk=round(physics_risk, 3),
        physics_risk=round(physics_risk, 3),
        ml_delta=0.0,
        criterion_met=crit,
        dsv_today=dsv_today,
        dsv_accum=dsv_accum,
        wet_hours=days[0].wet_hours if days else 0,
        min_temp_c=days[0].min_temp_c if days else 0.0,
        mean_wet_temp_c=days[0].mean_wet_temp_c if days else 0.0,
    )
